### ISSUE TITLE
An updated encoding for ZSink

### ISSUE THREAD

**Issue by vasilmkd** [zio/zio#1605]
Let this issue serve as a revisit of #1470.

As you already know, the current encoding of `ZSink` is as follows:
```scala
trait ZSink[-R, +E, +A0, -A, +B] {
  type State
  // initial state with no leftover
  def initial: ZIO[R, E, Step[State, Nothing]]

  // produce a new state after consuming a stream element and potentially finish
  // along with producing a leftover of type A0
  def step(state: State, a: A): ZIO[R, E, Step[State, A0]]

  // finish with the production of a final value of type B
  def extract(state: State): ZIO[R, E, B]

  // if you want to get the Sink remainder you need to use
  // Step.leftover(step: Step[State, A0]): Chunk[A0]
}
```

We are currently in the process of getting rid of the `Step`, which is an effort towards ZIO 1.0 (#1112). Open issue is here #538 and WIP PR is here #1560.

This effort changes the encoding of `ZSink` to the following:
```scala
trait ZSink[-R, +E, +A0, -A, +B] {
  type State

  // initial state
  def initial: ZIO[R, E, State]

  def step(state: State, a: A): ZIO[R, E, State]

  def extract(state: State): ZIO[R, E, B]

  // Used to be a part of the `Step` API. Decide whether the Sink should
  // continue given the current state.
  def cont(state: State): Boolean

  // Used to be a part of the `Step` API. Return the produced remainder.
  def leftover(state: State): Chunk[A0]
}
```

However, the second encoding, without `Step` currently has a drawback. It cannot properly be used to define combinators which feed two underlying sinks in parallel and need to decide the final result of the combined sink, as well as the remainder that is to be produced. I'll use the `orElse` combinator to illustrate.

`orElse` is a combinator that feeds the stream elements to both of the underlying sinks. The left sink is given an advantage such that the whole sink succeeds if the left sink succeeds, and then the left result and remainder are produced. The combined sink also continues for as long as the left sink needs to continue. If it happens that the right sink is finished and successfully extracted along with the remainder, it continues to accumulate the incoming stream elements in its remainder, in case the left sink fails in the future. In that case, the result of the right sink and its remainder will be produced as the result of the combined sink. This allows the `orElse` combinator to function as the `choice` operator in parser combinator libraries.

To illustrate why the encoding is insufficient, let's suppose that we have a combined sink of `ZSink.collectAll[Int]` on the LHS and `ZSink.collectAllN[Int](2)` on the RHS of the combined sink. Both sinks accumulate incoming stream elements into lists, however, the sink on the RHS stops after producing a `List` of exactly 2 elements. Any other passed stream element would be accumulated in it's remainder. The sink on the LHS is peculiar because it is a sink that always signals that it accepts new stream elements, but it will also never fail when extracted, producing a `List` of arbitrary length. That means that the sink on the LHS can be in this state, neither fully extracted with a produced remainder, nor we are sure that it will not fail when extracted (of course, the `orElse` combinator works on arbitrary streams).

So, let's say that we have a stream of 5 elements and all 5 have already passed through the sink. The state of the sink at this point is as follows: on the LHS, we have a sink that is ready to consume more elements (thus the combined sink is not fully done) and on the RHS we have a sink that is fully extracted, produced a list of 2 elements and accumulated a remainder `Chunk` of 3 elements. Because the stream is done (let's suppose that we have called `run` on the stream with the combined sink), it's time to extract the combined sink and produce a value.

At extraction time, the combined `orElse` sink needs to honor its contract and try to extract the LHS sink first. Because it will not fail, it will produce the full `List` of all 5 stream elements. However, the state of the sink will not change, because `def extract(state: State): ZIO[R, E, B]` does not alter the state of the sink, it just produces a value. Feeding that same state to the `def leftover(state: State): Chunk[A0]` method in order to obtain the sink remainder is now where the problem lies.

Now sure, we can say, the LHS is not signalling that it's finished, therefore we should return an empty `Chunk` in that case. However, what if the extraction of the LHS side failed, and therefore the true result of the combined sink is the result and the remainder of the RHS sink. There is no way to encode this information back into the state (because `extract` does not modify the state, and it shouldn't).

When we redid the underlying implementation of `ZStream`, @iravid experimented with a radically new encoding for `ZSink` to go along. However, as documented in #1470, he showed that there is not much to be gained from it. I do, however think that there is some merit to the encoding expressed there, in that it produced a tuple of both the sink result and the sink remainder. Therefore, I believe that the encoding of `ZSink` should be changed to the following:

```scala
trait ZSink[-R, +E, +A0, -A, +B] {
  type State

  def initial: ZIO[R, E, State]

  def step(state: State, a: A): ZIO[R, E, State]

  // return the result and the remainder in one go.
  def extract(state: State): ZIO[R, E, (B, Chunk[A0])]

  def cont(state: State): Boolean

  // no need for a leftover method with confusing semantics
  // because there is no way to encode in the state that the
  // sink has been successfully extracted
}
```

Going back to the `orElse` discussion, the sink that successfully produces a value also gets to return its remainder correctly, which is the intended behavior of the combined sink.

I thank in advance anyone that takes the time to read through this, hopefully it makes sense. Please ask if further clarification is needed, I understand that this may be hard to read without the proper context and implementation details.

I'm looking forward to a discussion.

cc @jdegoes

**Comment by iravid** [comment]
That's very well written! I'm onboard with the change. I thought this problem could be solved by encoding more information in the internal sink state, but did not consider the problem of sinks that always return `more`.

**Comment by ghostdogpr** [comment]
Closed by #1560
